### PR TITLE
[IR] Create documentation for tensors

### PR DESCRIPTION
**Pull request by justinchuby**
Document the TensorProtocol and the various tensor classes. When we create a helper function to create tensors in the IR, we should document that as well.

### CODE REVIEW

**Review by thiagocrepaldi — APPROVED** [review]

File: docs/intermediate_representation/tensors.md [hunk 1]
```diff
@@ -0,0 +1,154 @@
+# Tensor representation in the IR
+
+The ONNX IR offers the :class:`TensorProtocol` interface for usings different data structures as backing data for tensors. Besides the traditional ``onnx.TensorProto``, you can also use ``np.ndarray``, ``torch.Tensor``, and virtually anything else to represent tensors in the graph. This allows for  They can be accessed and serialized via the same `TensorProtocol` interface.
```
Comment (gramalingam) [inline docs/intermediate_representation/tensors.md:3]: change "They can be" to "them to be"?

File: docs/intermediate_representation/tensors.md [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+    tensor = ir.TensorProtoTensor(tensor_proto)
+    print(tensor: ", tensor)  # TensorProtoTensor<INT16,[3]>(name='tensor')
+    print("shape: ", tensor.shape)  # ir.Shape([3])
+    print("dtype: ", tensor.dtype)  # ir.DataType.INT16
+    print(tensor.raw == tensor_proto)  # The raw field is the exact tensor_proto provided at initialization
+    print("tobytes: ", tensor.tobytes())  # b'\x01\x00\x02\x00\x03\x00'
+    print("numpy: ", tensor.numpy())  # array([1, 2, 3], dtype=int16)
+```
+
+When tensors store their data in a different data field like `int32_data`, use the corresponding class as a wrapper:
```
Comment (gramalingam) [inline docs/intermediate_representation/tensors.md:36]: Unclear. This looks like the same wrapper as the one above? But the description makes it sound like it is a different wrapper?

File: docs/intermediate_representation/tensors.md [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+
+```{eval-rst}
+.. exec_code::
+    import onnx
+    from onnxscript import ir
+
+    tensor_proto = onnx.helper.make_tensor(
+        "tensor", onnx.TensorProto.INT16, (3,), b'\x01\x00\x02\x00\x03\x00', raw=True
+    )
+    tensor = ir.TensorProtoTensor(tensor_proto)
```
Comment (gramalingam) [inline docs/intermediate_representation/tensors.md:27]: Does the API allow users to create this tensor without using `onnx.helper.make_tensor`? How would they do that?

When users create the IR representation from a pre-existing tensor-proto, ideally they should call a single API method (without having to handle the different cases of a TensorProto themselves).

Comment (justinchuby) [reply docs/intermediate_representation/tensors.md:27]: Good point - I am going to combine all tensor proto classes and simplify as part of https://github.com/microsoft/onnxscript/pull/1441

Comment (justinchuby) [reply docs/intermediate_representation/tensors.md:27]: Done: https://github.com/microsoft/onnxscript/pull/1441/commits/040827821422f192cf2febccfbdd890ea7c60df7

Comment (justinchuby) [reply docs/intermediate_representation/tensors.md:27]: > Does the API allow users to create this tensor without using onnx.helper.make_tensor? 

Yes - all users need is a tensor proto as input

I expect to create a similar (make_tensor) api that will produce ir tensor for us

File: docs/intermediate_representation/tensors.md (hunk 2 above)
Comment (justinchuby) [reply docs/intermediate_representation/tensors.md:36]: Removed

File: docs/intermediate_representation/tensors.md (hunk 1 above)
Comment (justinchuby) [reply docs/intermediate_representation/tensors.md:3]: Thanks for the catch. That was incomplete 😅

**Review by gramalingam — APPROVED** [review]

**Comment by justinchuby** [comment]
1. single backtick in markdown
2. {py:class}`ir.Value <onnxscript.ir._core.Value>`

**Pull request merged**